lib/pure/strinterp.nim

```diff
@@ -0,0 +1,320 @@
+  ## Converts ``n`` to string. If ``n`` is `SomeReal`, it casts to `int64`.
+  ## Conversion is done using ``radix``. If result's length is lesser then
+  ## ``len``, it aligns result to the right with ``fill`` char.
+  ## If ``len`` is negative, the result is aligned to the left.
+  ## If `lowerCase` is true, formatted string will be in the lower case.
+  when n is SomeUnsignedInt:
+    var v = n.uint64
+    let s = false
+  else:
+    var v = n.int64
+    let s = v.int64 < 0
+    if s:
+      v = v * -1
+
+  if v == 0:
+    result = "0"
+  else:
+    result = ""
+    while v > (type(v))0:
+      let d = v mod (type(v))radix
+      v = v div (type(v))radix
+      result.add(mkDigit(d.int, lowerCase))
+    for idx in 0..<(result.len div 2):
+      swap result[idx], result[result.len - idx - 1]
+  var length = abs(len)
+  if length == 0 or (s and (result.len >= length - 1)) or (not s and (result.len >= length)):
+    if s:
+      result = "-" & result
+  elif len < 0:
+    if s:
+      result = "-" & result
+    for i in result.len..<length:
+      result.add(fill)
+  else:
+    if fill != '0':
+      # The sign must be near the number
+      if s:
+        result = "-" & result
+    var toFill = length - result.len
+    var prefix = newString(toFill)
+    for idx in 0..<toFill:
+      prefix[idx] = fill
+    if fill == '0' and s:
+      prefix[0] = '-'
+    result = prefix & result
+
+proc alignStr(s: string, len: int, fill = ' ', trunc = false): string =
+  ## Aligns ``s`` using ``fill`` char to the right if ``len`` is
+  ## positive, or to the left, if ``len`` is negative.
+  ## If the length of ``s`` is bigger then `abs(len)` and ``trunc`` == true,
+  ## truncates ``s``
+  if len == 0:
+    result = s
+  elif trunc and s.len > abs(len):
+    result = s
+    result.setLen(abs(len))
+  else:
+    let fillLength = abs(len) - s.len
+    if fillLength <= 0:
+      result = s
+    elif len > 0:
+      result = repeat(fill, fillLength) & s
+    else:
+      result = s & repeat(fill, fillLength)
+
+proc floatToStr(v: SomeNumber, len = 0, prec = 0, sep = '.', fill = ' ',  scientific = false): string =
+  ## Converts ``v`` to string with precision == ``prec``. If result's length
+  ## is lesser then ``len``, it aligns result to the right with ``fill`` char.
+  ## If ``len`` is negative, the result is aligned to the left.
+  let f = if scientific: ffScientific else: if prec == 0: ffDefault else: ffDecimal
+  if len > 0 and v < 0 and fill == '0':
+    result = "-" & alignStr(formatBiggestFloat(-v.BiggestFloat, f, prec, sep), len-1, fill)
+  else:
+    result = alignStr(formatBiggestFloat(v.BiggestFloat, f, prec, sep), len, fill)
+
+# -----------------------------------------------------------------------------
+# nimboost's richstring
+# -----------------------------------------------------------------------------
+
+proc parseIntFmt(fmtp: string): tuple[maxLen: int, fillChar: char] =
```

Comment:
Would be nice to see some comments with examples of what this parses.

Same for the functions below.

Reply:
I have added some docstring, but note that this is not the best place to document the behavior for users, because these are only the internal helper functions. From a user perspective this is just about getting the traditional printf like behavior: the pad symbol of normal formatters like `%5d` are a plain space, and `%05d` is a special case, where the pad symbol becomes a zero.

Reply:
This isn't for users, it's for the programmer. It's nice to see an example of what this procedure parses.

Reply:
Are you happy with the comments I have added? Or anything else to review? (I didn't move it into strutils now, as wished by Araq)